### PR TITLE
add auto scale module

### PR DESCRIPTION
**Pull request by yuwzho**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

azure_rm_autoscale

##### ANSIBLE VERSION

```
2.7
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        description:
+            - The name of the resource group.
+        required: True
+    name:
+        description:
+            - The name of the Auto Scale Setting.
+    format:
+        description:
+            - Format of the data returned.
+            - If C(raw) is selected information will be returned in raw format from Azure Python SDK.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py]: just remove C(raw) as we will no longer do it

File: test/integration/targets/azure_rm_autoscale/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    format: curated
+  register: facts
+
+- name: use facts to send param to auto scale
+  azure_rm_autoscale:
+    location: "{{ facts.autoscales[0].location }}"
+    name: "{{ facts.autoscales[0].name }}"
+    notifications: "{{ facts.autoscales[0].notifications }}"
```
Comment (zikalino) [inline test/integration/targets/azure_rm_autoscale/tasks/main.yml]: looks great! 👍

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+---
+module: azure_rm_autoscale
+version_added: "2.7"
+short_description: Manage Azure image.
+description:
+    - Create, delete an image from virtual machine, blob uri, managed disk or snapshot.
+options:
+    target:
+        description:
+        - The resource identifier of the resource that the autoscale setting should be added to.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: I think just "The identifier of the resource to apply autoscale setting"

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+        - It can be the resource id of the resource.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: I would write: "it could be resource id string or a dictionary containing C(name)....."

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+    resource_group:
+        required: true
+        description: resource group of the resource.
+    enabled:
+        type: bool
+        description: Specifies whether automatic scaling is enabled for the resource.
+        default: true
+    profiles:
+        description:
+        - the collection of automatic scaling profiles that specify different scaling parameters for different time periods.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: Uppercase "The"

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+                required: true
+                description:
+                - The number of instances that will be set if metrics are not available for evaluation.
+                - The default is only used if the current instance count is lower than the default.
+            min_count:
+                description: the minimum number of instances for the resource.
+            max_count:
+                description: the maximum number of instances for the resource.
+            recurrence_frequency:
+                default: None
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py:54]: i am a bit concerned about None, as it will interfere with Python's None

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 7]
```diff
@@ -0,0 +1,620 @@
...
+                    value:
+                        description:
+                        - The number of instances that are involved in the scaling action.
+                        - This value must be 1 or greater.
+                    operator:
+                        default: GreaterThan
+                        description: The operator that is used to compare the metric data and the threshold.
+                        choices:
+                        - Equals
+                        - NotEquals
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py:135]: i think we should have these choices in python case perhaps "not_equals" etc....?
that applies to all options.

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 8]
```diff
@@ -0,0 +1,620 @@
...
+        - '6'
+      notifications:
+      - email_admin: True
+        email_co_admin: False
+        custom_emails:
+        - yuwzho@microsoft.com
+      name: scale
+      resource_group: foo
+
+- name: Delete an image
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: image?

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py (hunk 7 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_autoscale.py:135]: Actually, since the default is str(None), the Null can be convert as str(None)

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      name: auto_scale_name
+
+  - name: List instances of Auto Scale Setting
+    azure_rm_autoscale_facts:
+      resource_group: resource_group_name
+'''
+
+RETURN = '''
+azure_autoscale:
+    description: List of Azure Scale Settings dicts.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py:55]: real sample?

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 10]
```diff
@@ -0,0 +1,620 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_autoscale
+version_added: "2.7"
+short_description: Manage Azure image.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: update image

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+description:
+    - Create, delete an image from virtual machine, blob uri, managed disk or snapshot.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: update description and short_description

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 12]
```diff
@@ -0,0 +1,620 @@
...
+                - Second
+                - Minute
+                - Hour
+                - Day
+                - Week
+                - Month
+                - Year
+            recurrence_timezone:
+                description:
+                - the timezone of repeating times at which this profile begins.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: aligning first char case sensitive

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py [hunk 13]
```diff
@@ -0,0 +1,620 @@
...
+    def get_auto_scale(self):
+        try:
+            return self.monitor_client.autoscale_settings.get(self.resource_group, self.name)
+        except Exception as exc:
+            self.log('Error: failed to get auto scale settings {0} - {1}'.format(self.name, str(exc)))
+            return None
+
+    def create_or_update_auto_scale(self, param):
+        try:
+            return self.monitor_client.autoscale_settings.create_or_update(self.resource_group, self.name, param)
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py:604]: why not to add auto_scale_to_doct in this method. caller doesn't need to care about refine output and how to map then.

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale.py (hunk 11 above)
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale.py]: call  out it's  verified with vmss only for now?

**Comment by yuwzho** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@yuwzho has finished change this PR, @yungezz @zikalino Could you help review this PR again? Thanks!

**Comment by Fred-sun** [comment]
@yungezz @zikalino could help review this PR? thanks！

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            self.log('Could not get facts for autoscale {0} - {1}.'.format(self.name, str(e)))
+        return result
+
+    def list_by_resource_group(self):
+        results = []
+        try:
+            response = self.monitor_client.autoscale_settings.list_by_resource_group(self.resource_group)
+            results = [auto_scale_to_dict(item) for item in response if self.has_tags(item.tags, self.tags)]
+        except CloudError as e:
+            self.log('Could not get facts for autoscale {0} - {1}.'.format(self.name, str(e)))
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py]: add correlation id from e.request_id

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_autoscale_facts
+version_added: "2.7"
+short_description: Get Auto Scale Setting facts.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py]: Azure

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+  - name: List instances of Auto Scale Setting
+    azure_rm_autoscale_facts:
+      resource_group: resource_group_name
+'''
+
+RETURN = '''
+azure_autoscale:
+    description: List of Azure Scale Settings dicts.
+    returned: always
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py:56]: detail return fields

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+            self.results['autoscales'] = self.list_by_resource_group()
+        return self.results
+
+    def get(self):
+        result = []
+        try:
+            instance = self.monitor_client.autoscale_settings.get(self.resource_group, self.name)
+            result = [auto_scale_to_dict(instance)]
+        except CloudError as e:
+            self.log('Could not get facts for autoscale {0} - {1}.'.format(self.name, str(e)))
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py]: add e.request_id as correlation id

File: lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py [hunk 18]
```diff
@@ -158,7 +158,7 @@ class AzureRMStorageAccount(AzureRMModuleBase):
     def __init__(self):
 
         self.module_arg_spec = dict(
-            account_type=dict(type='str', choices=[], aliases=['type']),
+            account_type=dict(type='str', choices=['Premium_LRS', 'Standard_GRS', 'Standard_LRS', 'Standard_RAGRS', 'Standard_ZRS'], aliases=['type']),
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_storageaccount.py]: unrelated change?

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+  - name: List instances of Auto Scale Setting
+    azure_rm_autoscale_facts:
+      resource_group: resource_group_name
+'''
+
+RETURN = '''
+azure_autoscale:
+    description: List of Azure Scale Settings dicts.
+    returned: always
+    type: list
```
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py:57]: yes, def needs a real sample return value

**Review by nitzmahone — CHANGES_REQUESTED** [review]
just the question about the storageaccount change, and documenting the return value

File: lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py (hunk 17 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_autoscale_facts.py]: Unfortunately, this exception is not kind of cloud error, but azure.mgmt.monitor.models.ErrorResponse. It only contains code and message.

**Comment by yuwzho** [comment]
@mattclay I think the storageaccount issue should be solved, but I have no idea why the sanity test failed here

**Comment by yuwzho** [comment]
ready_for_review

@mattclay Would you mind sharing the status of StorageAccount?

**Comment by yuwzho** [comment]
@mattclay

**Comment by Fred-sun** [comment]
@yungezz @zikalino @nitzmahone  Could you help review this PR again, It has ready for review again. Thanks!

**Comment by yuwzho** [comment]
@mattclay I add an idempotent new module which add a client in the azure_rm_common, but it triggers all test and results in timeout. How can I solve it?

**Review by nitzmahone — APPROVED** [review]

**Comment by nitzmahone** [comment]
The timeouts aren't the fault of this module; we'll have to figure it out anyway.
